### PR TITLE
Fixes #1464 throw when running the same twice

### LINKED ISSUE #1464: Throw exception when running the same simulation with runSimulations
as we are caching with id

### PR DESCRIPTION
**Pull request by msevestre**

### CODE REVIEW

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs [hunk 1]
```diff
@@ -38,56 +38,99 @@ public interface IConcurrencyManager
...
+      /// </summary>
+      /// <typeparam name="TData">Data type to consume by the worker function</typeparam>
+      /// <param name="data">List of data to consume by the workers</param>
+      /// <param name="action">
+      ///    An action running on each worker on each piece of data.
+      /// </param>
+      /// <param name="cancellationToken">Cancellation token to cancel the threads</param>
+      /// <param name="numberOfCoresToUse">Number of cores to use. Use null to take all cores</param>
+      /// <returns>Dictionary binding a result for each input data after running the action on it</returns>
+      Task RunAsync<TData>(
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:64]: @abdelr As discussed. Thanks for the suggestion

File: src/OSPSuite.R/Domain/ConcurrentRunSimulationBatch.cs [hunk 2]
```diff
@@ -41,12 +41,9 @@ internal SimulationBatch AddNewBatch()
 
       public IReadOnlyCollection<SimulationBatch> SimulationBatches => _simulationBatches;
 
-      public string AddSimulationBatchRunValues(SimulationBatchRunValues simulationBatchRunValues)
+      public void AddSimulationBatchRunValues(SimulationBatchRunValues simulationBatchRunValues)
       {
-         var id = generateId();
-         simulationBatchRunValues.Id = id;
```
Comment (msevestre) [inline src/OSPSuite.R/Domain/ConcurrentRunSimulationBatch.cs:47]: I am moving the creation of the Id to the object itselft...

File: src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs [hunk 3]
```diff
@@ -98,22 +98,23 @@ public void Clear()
...
-               settings => Enumerable.Range(0, settings.MissingBatchesCount).Select(_ => settings)
-            ).ToList(),
-            (settings, ct) => settings.AddNewBatch(),
+            batch => Enumerable.Range(0, batch.MissingBatchesCount).Select(_ => batch)
+         ).ToList();
+
+         return _concurrencyManager.RunAsync(
+            batches,
+            //Use {} to specify using the overload without return value
+            (batch, ct) => { batch.AddNewBatch(); },
```
Comment (msevestre) [inline src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs:115]: This is the only downside of using the same name... I need to use the {} to that it picks the right overload

File: tests/OSPSuite.Presentation.Tests/Services/DataImporterSpecs.cs [hunk 4]
```diff
@@ -186,19 +189,30 @@ public void should_import_simple_data_from_csv()
       }
 
       [Observation]
-      public void should_return_empty_on_invalid_file_name()
+      [TestCase("IntegrationSample1.csv", 2)]
+      [TestCase("sample_non_existent.csv", 2)]
```
Comment (msevestre) [inline tests/OSPSuite.Presentation.Tests/Services/DataImporterSpecs.cs:193]: oups... this does not belong here!

File: tests/OSPSuite.R.Tests/Services/ConcurrentSimulationRunnerSpecs.cs [hunk 5]
```diff
@@ -235,11 +238,18 @@ public override void GlobalContext()
...
+         {
+            ParameterValues = new[] {3.5, 0.53}
+         };
+         _parValues3 = new SimulationBatchRunValues
+         {
+            ParameterValues = new[] {3.5, 0.53}
+         };
 
          _simulationBatch1.AddSimulationBatchRunValues(_parValues1);
-         _simulationBatch2.AddSimulationBatchRunValues(_parValues1);
```
Comment (msevestre) [inline tests/OSPSuite.R.Tests/Services/ConcurrentSimulationRunnerSpecs.cs:241]: Adding the same object multiple time is not allowed anymore... unique ids

**Review by msevestre — COMMENTED** [review]
@abdelr Ready to review

File: src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs [hunk 6]
```diff
@@ -38,56 +38,99 @@ public interface IConcurrencyManager
...
                {
-                  cancellationToken.ThrowIfCancellationRequested();
-                  try
-                  {
-                     results.TryAdd(
-                        datum, 
-                        new ConcurrencyManagerResult<TResult>(datum.Id, action(datum, cancellationToken))
-                     );
-                  }
-                  catch (Exception e)
```
Comment (abdelr) [inline src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:79]: You removed the try/catch. This was here to make sure that if one run fails (running one simulation throws an exception for example) the rest are still recorded. Like this I think you would not be able to return anything at all. Right?

Comment (msevestre) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:79]: yep. But here we have no way of knowing which one failed as we have no identifier.

Comment (abdelr) [reply src/OSPSuite.Core/Domain/Services/ConcurrencyManager.cs:79]: Fair enough.

**Review by abdelr — APPROVED** [review]

**Pull request merged**